### PR TITLE
FEAT: Permissions

### PR DESCRIPTION
**Pull request by naftalima**
## Objective and Motivation
- Only authenticated users can access endpoints
- A custom permission class `IsTheTrainerOfTheTeam` for only the trainer of the team can select pokemons for that team.

### CODE REVIEW

File: backend/battles/api/permissions.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+    message = "This user isn't the trainer of this team."
+
+    def has_permission(self, request, view):
+        if request.user.is_authenticated:
+            return True
+        return False
+
+    def has_object_permission(self, request, view, obj):
+        if str(obj.trainer) == str(request.user):
```
Comment (rsarai) [inline backend/battles/api/permissions.py:14]: I advise you to be more specific here: `if obj.trainer.email == request.user.email:`

File: backend/battles/api/permissions.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+from rest_framework import permissions
+
+
+class IsTheTrainerOfTheTeam(permissions.BasePermission):
```
Comment (rsarai) [inline backend/battles/api/permissions.py:4]: You don't need to add the `The` word

File: backend/battles/api/tests/test_views.py [hunk 3]
```diff
@@ -156,3 +156,59 @@ def side_effect_func(pokemon_name):
...
+                    "attack": 40,
+                    "hp": 20,
+                    "name": "nidorina",
+                    "img_url": "https://raw.githubusercontent.com"
+                    "/PokeAPI/sprites/master/sprites/pokemon/25.png",
+                    "poke_id": 10,
+                }
+            return fake_json
+
+        mock_get_pokemon.side_effect = side_effect_func
```
Comment (rsarai) [inline backend/battles/api/tests/test_views.py:198]: Please check if this is being called

File: backend/battles/api/tests/test_views.py [hunk 4]
```diff
@@ -156,3 +156,59 @@ def side_effect_func(pokemon_name):
...
+        team_pokemon_data = {
+            "pokemon_1": "pikachu",
+            "position_1": 1,
+            "pokemon_2": "pikachu",
+            "position_2": 2,
+            "pokemon_3": "pikachu",
+            "position_3": 3,
+        }
+
+        self.auth_client.patch(
```
Comment (rsarai) [inline backend/battles/api/tests/test_views.py:209]: The permission should raise a different status_code, please check for it

**Review by rsarai — CHANGES_REQUESTED** [review]

File: backend/battles/api/permissions.py (hunk 2 above)
Comment (naftalima) [reply backend/battles/api/permissions.py]: it was really weird, thank you.

File: backend/battles/api/permissions.py (hunk 1 above)
Comment (naftalima) [reply backend/battles/api/permissions.py:14]: done!

File: backend/battles/api/tests/test_views.py (hunk 3 above)
Comment (naftalima) [reply backend/battles/api/tests/test_views.py:198]: is to check if the response is 403 (forbidden) right?

File: backend/battles/api/tests/test_views.py (hunk 4 above)
Comment (naftalima) [reply backend/battles/api/tests/test_views.py]: I added the check if the answer is 403

File: backend/battles/api/tests/test_views.py (hunk 3 above)
Comment (naftalima) [reply backend/battles/api/tests/test_views.py:198]: Sorry, but I didn't understand what's not being called.

Comment (rsarai) [reply backend/battles/api/tests/test_views.py]: If the user can't access view, this mock will never be called. You should add: `mock_get_pokemon.assert_not_called()`

**Review by rsarai — APPROVED** [review]

**Comment by rsarai** [comment]
Please fix the mock before merging it :eyes:

**Pull request merged**